### ISSUE TITLE
PDFSharp v6.2 not FIPS compliant

### ISSUE THREAD

**Issue by ketchup201** [empira/PDFsharp#205]
I don't think this requires sample code, as it's evident in the source.

```
using (var fs = File.OpenRead(filename))
{
    PdfPage pdfPage = outputPdf.AddPage();
    pdfPage.Width = XUnit.FromMillimeter(mmX);
    pdfPage.Height = XUnit.FromMillimeter(mmY);
    pdfPage.Orientation = PdfSharp.PageOrientation.Portrait;

    using (var g = XGraphics.FromPdfPage(pdfPage))
    {
        using (XImage image = XImage.FromStream(fs))
        {
            g.DrawImage(image, 0, 0, pdfPage.Width.Point, pdfPage.Height.Point);
        }
    }
}

```
This throws an error in DrawImage. The error is that a non-FIPS compliant method is being used. The PdfImageTable:ImageSelector class has a constructor that makes heavy use of System.Security.Cryptography.MD5.Create().

This method is known to be unsafe and even though we're only using it to compute hashes, this will crash on any machine where FIPS compliance is required.

**Comment by ThomasHoevel** [comment]
> I don't think this requires sample code, as it's evident in the source.
> [...]
> This method is known to be unsafe and even though we're only using it to compute hashes, this will crash on any machine where FIPS compliance is required.

A link about how to configure a machine for FIPS compliance and links to information about FIPS compliance cannot harm.

PDFsharp has to use MD5 when opening PDF files that are secured with MD5, so we cannot eliminate it everywhere without impacting functionality.
And there is no benefit in using a better, more secure, and slower hashing method in ImageSelector.

**Comment by ketchup201** [comment]
Thanks for the reply.. Here's a background on FIPS and Microsoft cryptography:
https://learn.microsoft.com/en-us/windows/security/security-foundations/certification/fips-140-validation

The issue here is that in production machines that have FIPS enabled, any MD5 algorithm throws an exception. Here's a link describing how to turn it on:
https://learn.microsoft.com/en-us/answers/questions/1188777/fips-for-windows-11

If MD5 is required for PDF, they may want to consider implementing their own MD5 hashing algorithm without the use of the Microsoft one, to work around the issue of PDFSharp not being usable in more secure environments.

**Comment by StLange** [comment]
Intersting issue. Thomas has already changed the hash algorithm used for **ImageSelector**.

We still need MD5 in PDFsharp for opening old PDF files that based on MD5 hash codes. But we will replace the .NET MD5 invocation by code we used many years ago for the PDFsharp Silverlight build. This code  based on the class **MD5Managed** by two Microsoft engineers. You can still find this code, e.g. here: [MD5](https://github.com/microsoft/surveillance-camera/blob/master/ServerApi/SurveillanceCamera/SurveillanceCamera/MD5.cs)

This code is fast and correct and comes to PDFsharp in the next preview we will release early in December.

This should fix your issues with FIPS.

**Comment by ThomasHoevel** [comment]
Issue should be resolved with PDFsharp 6.2.0 Preview 2.
